### PR TITLE
Feature initial staging

### PR DESCRIPTION
**Pull request by g-walley**
- Added initial-staging to StagedTree
- Fixed docstring for prior param in AHC method
- Added validation methods for initial staging

### CODE REVIEW

**Comment by jackcarter23** [comment]
Noticed a small bug but I think I've lost my access to fix it!  Could you change the line
def _validate_hyperstage(initial_staging, hyperstage):
to
def _validate_hyperstage(self,initial_staging, hyperstage):

**Comment by jackcarter23** [comment]
Also I have found that by using the initial stage along with the hyperstage arguments, you are able to make any ST you like (not restricted to the AHC output).  A nice little trick it might be worth pointing out in the documentation so that people can make completely custom stagings

**Comment by g-walley** [comment]
> Also I have found that by using the initial stage along with the hyperstage arguments, you are able to make any ST you like (not restricted to the AHC output). A nice little trick it might be worth pointing out in the documentation so that people can make completely custom stagings

Ok, I can have a look at doing that :)

File: src/cegpy/trees/_staged.py [hunk 1]
```diff
@@ -518,7 +545,9 @@ def _independent_hyperstage_generator(hyperstages: List[List]) -> List[List[List
 
         return new_hyperstages
 
-    def _execute_ahc(self, hyperstage=None) -> Tuple[List, float, List]:
+    def _execute_ahc(
+        self, hyperstage=None, initial_staging=Optional[List[List[str]]]
+    ) -> Tuple[List, float, List]:
```
Comment (ashenvi10) [inline src/cegpy/trees/_staged.py:550]:
```suggestion
    def _execute_ahc(
        self, hyperstage: Optional[List[List[str]]] = None, initial_staging: Optional[List[List[str]]] = None
    ) -> Tuple[List, float, List]:
```

File: src/cegpy/trees/_staged.py [hunk 2]
```diff
@@ -646,16 +709,40 @@ def _generate_colours_for_situations(self, merged_situations, colour_list):
                 except KeyError:
                     self.nodes[node]["colour"] = "white"
 
+    @staticmethod
+    def _validate_hyperstage(initial_staging: List, hyperstage):
```
Comment (ashenvi10) [inline src/cegpy/trees/_staged.py:713]:
```suggestion
    def _validate_hyperstage(initial_staging: List, hyperstage) -> bool:
```

File: src/cegpy/trees/_staged.py [hunk 3]
```diff
@@ -646,16 +709,40 @@ def _generate_colours_for_situations(self, merged_situations, colour_list):
...
+            # init_stage is a subset of only one stage in hyperstages
+            for h_stage in hyperstage:
+                if set(init_stage).issubset(h_stage):
+                    found += 1
+            if found == 0:
+                return False
+        return True
+
+    @staticmethod
+    def _validate_initial_staging_mutually_exclusive(initial_staging: List):
```
Comment (ashenvi10) [inline src/cegpy/trees/_staged.py:725]:
```suggestion
    def _validate_initial_staging_mutually_exclusive(initial_staging: List) -> bool:
```

File: src/cegpy/trees/_staged.py [hunk 4]
```diff
@@ -673,9 +760,11 @@ def calculate_AHC_transitions(
         :rtype: Dict
         """
 
-        self._store_params(prior, alpha, hyperstage)
+        self._store_params(prior, alpha, hyperstage, initial_staging)
 
-        loglikelihood, merged_situations = self._execute_ahc()
+        loglikelihood, merged_situations = self._execute_ahc(
+            initial_staging=initial_staging
+        )
```
Comment (ashenvi10) [inline src/cegpy/trees/_staged.py:767]: Why don't we pass `hyperstage` as an argument here?

**Review by ashenvi10 — APPROVED** [review]
Some minor comments, otherwise LGTM

File: src/cegpy/trees/_staged.py (hunk 4 above)
Comment (g-walley) [reply src/cegpy/trees/_staged.py:767]: For some daft reason, you can pass it in as an argument, or it is pulled directly from `self`.
It is written to `self.hyperstage` in `self._store_params` so it should work as is. I think there needs to be some general packaging cleaning up. But that is out of scope for this PR.

**Pull request merged**